### PR TITLE
addressing #2488-H3-indexes-bottom-up

### PR DESCRIPTION
**Pull request by haitamlaarabi**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2586)

### CODE REVIEW

File: src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    )
+    val services = injector.getInstance(classOf[BeamServices])
+    DefaultPopulationAdjustment(services).update(scenario)
+    val demandPoints = scenario.getPopulation.getPersons.asScala
+      .flatMap(_._2.getPlans.get(0).getPlanElements.asScala.filter(_.isInstanceOf[Activity]))
+      .map(_.asInstanceOf[Activity].getCoord)
+      .toArray
+    val indexes = services.beamScenario.h3taz.getDemandInferredH3IndexSet(demandPoints, 10, 6, 10)
+    assert(indexes.nonEmpty, "Set of H3 indexes should not be empty.")
+    assert(indexes.flatMap(_._2).length == demandPoints.length, "Not all coordinates were matched to an H3 Index")
```
Comment (wrashid) [inline src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala:57]: please add assert for indexes.map(x=>x._2.length).max<=maxNumberOfDemandPointsPerHex (just found that this is failing at the moment).

File: src/main/scala/beam/agentsim/infrastructure/taz/H3TAZ.scala [hunk 2]
```diff
@@ -56,6 +57,29 @@ case class H3TAZ(network: Network, tazTreeMap: TAZTreeMap, beamConfig: BeamConfi
   def getIndex(c: Coord): HexIndex = getIndex(c, resolution)
   def getResolution: Int = resolution
 
+  // inputs: lowestLevel(=startingLevel), maxNumberOfDemandPointsPerHex, maxLevel, coordinates of demands
+  //
+  //Outputs: h3Indexes at different resolution levels
+  def getDemandInferredH3IndexSet(
+    demandPoints: Array[Coord],
+    maxNumberOfDemandPointsPerHex: Int,
+    lowestResolution: Int,
```
Comment (wrashid) [inline src/main/scala/beam/agentsim/infrastructure/taz/H3TAZ.scala:66]: if the problem is over constrained, throw exception, e.g. no solution found

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

File: src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    val highestResolution = 9
+    val indexes =
+      H3TAZ.getDataPointsInferredH3IndexSet(demandPoints, maxDemandPoints, lowestResolution, highestResolution)
+    assert(indexes.nonEmpty, "Set of H3 indexes should not be empty.")
+    assert(indexes.map(_._2.length).sum == demandPoints.length, "Not all coordinates were matched to an H3 Index")
+    assert(
+      indexes.map(_._2.length).max <= maxDemandPoints,
+      "number of data points should be lower than or equal to 100"
+    )
+    assert(
```
Comment (wrashid) [inline src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala:74]: no need to test this

Comment (wrashid) [inline src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala:74]: how about we test that the output indexes are in the correct range (we could omit testing the config here):

// Check that all indexes are within constrain
indexes.foreach { case (h3Index, coords) =>
  val resolution = H3TAZ.H3.h3GetResolution(h3Index)
  require(resolution >= lowestResolution && resolution <= highestResolution, s"Expected to have resolution in [$lowestResolution, $highestResolution], but got $resolution")
}

for this to work you might have to make H3TAZ.H3 public or provide a getter

**Comment by haitamlaarabi** [comment]
Test!

File: src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    indexes.foreach {
+      case (h3Index, _) =>
+        val resolution = H3TAZ.getResolution(h3Index)
+        require(
+          resolution >= lowestResolution && resolution <= highestResolution,
+          s"Expected to have resolution in [$lowestResolution, $highestResolution], but got $resolution"
+        )
+    }
+
+    // the following tests are unnecessary
```
Comment (REASY) [inline src/test/scala/beam/agentsim/infrastructure/taz/H3TAZSpec.scala:80]: Let's remove unnecessary code so it will be easier for the readers.

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Comment by wrashid** [comment]
Test!

**Pull request merged**